File: scripts/apply-clickhouse.js
const { chromium } = require('playwright');
const path = require('path');
const fs = require('fs');

const JOB_URL = 'https://job-boards.greenhouse.io/clickhouse/jobs/5803692004';
const RESUME_PDF = '/home/user/Agents/profile/Hisham Abboud CV.pdf';
const SCREENSHOTS_DIR = '/home/user/Agents/output/screenshots';

const CANDIDATE = {
  firstName: 'Hisham',
  lastName: 'Abboud',
  email: 'hiaham123@hotmail.com',
  phone: '+31 06 4841 2838',
  linkedin: 'linkedin.com/in/hisham-abboud',
  github: 'github.com/Hishamabboud',
  location: 'Eindhoven, Netherlands',
};

const COVER_LETTER_TEXT = fs.readFileSync(
  '/home/user/Agents/output/cover-letters/clickhouse-cloud-engineer.md',
  'utf8'
);

async function screenshot(page, name) {
  const filePath = path.join(SCREENSHOTS_DIR, `clickhouse-${name}.png`);
  await page.screenshot({ path: filePath, fullPage: true });
  console.log(`Screenshot saved: ${filePath}`);
  return filePath;
}

async function run() {
  const browser = await chromium.launch({
    headless: true,
    args: ['--no-sandbox', '--disable-setuid-sandbox'],
  });

  const context = await browser.newContext();
  const page = await context.newPage();

  console.log('Navigating to job application page...');
  await page.goto(JOB_URL, { waitUntil: 'networkidle', timeout: 60000 });
  await screenshot(page, '01-job-page-loaded');

  console.log('Page title:', await page.title());

  // Wait for the form to appear
  await page.waitForSelector('form', { timeout: 30000 });
  await screenshot(page, '02-form-visible');

  // --- Fill First Name ---
  const firstNameField = await page.locator('input[name="first_name"], input[id*="first_name"], input[placeholder*="First"]').first();
  if (await firstNameField.count() > 0) {
    await firstNameField.fill(CANDIDATE.firstName);
    console.log('Filled first name');
  } else {
    console.log('WARNING: first name field not found');
  }

  // --- Fill Last Name ---
  const lastNameField = await page.locator('input[name="last_name"], input[id*="last_name"], input[placeholder*="Last"]').first();
  if (await lastNameField.count() > 0) {
    await lastNameField.fill(CANDIDATE.lastName);
    console.log('Filled last name');
  } else {
    console.log('WARNING: last name field not found');
  }

  // --- Fill Email ---
  const emailField = await page.locator('input[name="email"], input[type="email"]').first();
  if (await emailField.count() > 0) {
    await emailField.fill(CANDIDATE.email);
    console.log('Filled email');
  } else {
    console.log('WARNING: email field not found');
  }

  // --- Fill Phone ---
  const phoneField = await page.locator('input[name="phone"], input[type="tel"], input[id*="phone"]').first();
  if (await phoneField.count() > 0) {
    await phoneField.fill(CANDIDATE.phone);
    console.log('Filled phone');
  } else {
    console.log('WARNING: phone field not found');
  }

  await screenshot(page, '03-basic-fields-filled');

  // --- Upload Resume ---
  if (fs.existsSync(RESUME_PDF)) {
    const fileInput = await page.locator('input[type="file"]').first();
    if (await fileInput.count() > 0) {
      await fileInput.setInputFiles(RESUME_PDF);
      console.log('Resume uploaded');
      await page.waitForTimeout(2000);
      await screenshot(page, '04-resume-uploaded');
    } else {
      console.log('WARNING: file input not found for resume');
    }
  } else {
    console.log('WARNING: Resume PDF not found at', RESUME_PDF);
  }

  // --- Fill LinkedIn ---
  const linkedinField = await page.locator('input[name*="linkedin"], input[id*="linkedin"], input[placeholder*="LinkedIn"]').first();
  if (await linkedinField.count() > 0) {
    await linkedinField.fill(`https://${CANDIDATE.linkedin}`);
    console.log('Filled LinkedIn');
  } else {
    console.log('LinkedIn field not found (may be optional)');
  }

  // --- Fill GitHub ---
  const githubField = await page.locator('input[name*="github"], input[id*="github"], input[placeholder*="GitHub"]').first();
  if (await githubField.count() > 0) {
    await githubField.fill(`https://${CANDIDATE.github}`);
    console.log('Filled GitHub');
  } else {
    console.log('GitHub field not found (may be optional)');
  }

  // --- Fill Cover Letter (if text area exists) ---
  const coverLetterArea = await page.locator('textarea[name*="cover"], textarea[id*="cover"], textarea[placeholder*="cover"]').first();
  if (await coverLetterArea.count() > 0) {
    await coverLetterArea.fill(COVER_LETTER_TEXT);
    console.log('Filled cover letter text area');
  } else {
    console.log('Cover letter text area not found (may be file upload only)');
  }

  // --- Fill Location if asked ---
  const locationField = await page.locator('input[name*="location"], input[id*="location"], input[placeholder*="ocation"]').first();
  if (await locationField.count() > 0) {
    await locationField.fill(CANDIDATE.location);
    console.log('Filled location');
  }

  await screenshot(page, '05-all-fields-filled');

  // --- Handle Visa Sponsorship dropdown ---
  // Look for any dropdown asking about visa/work authorization
  const visaSelects = await page.locator('select').all();
  for (const sel of visaSelects) {
    const id = await sel.getAttribute('id') || '';
    const name = await sel.getAttribute('name') || '';
    if (id.toLowerCase().includes('visa') || name.toLowerCase().includes('visa') ||
        id.toLowerCase().includes('sponsor') || name.toLowerCase().includes('sponsor') ||
        id.toLowerCase().includes('auth') || name.toLowerCase().includes('auth')) {
      // Select "No" for sponsorship (candidate is in Netherlands, presumably authorized)
      try {
        await sel.selectOption({ label: 'No' });
        console.log(`Set visa/sponsorship dropdown (${id || name}) to No`);
      } catch (e) {
        console.log('Could not set visa dropdown:', e.message);
      }
    }
  }

  await screenshot(page, '06-before-submit');

  // --- Log all form fields for debugging ---
  const allInputs = await page.locator('input, select, textarea').all();
  console.log(`\nTotal form elements found: ${allInputs.length}`);
  for (const el of allInputs) {
    const tag = await el.evaluate(e => e.tagName);
    const type = await el.getAttribute('type') || '';
    const name = await el.getAttribute('name') || '';
    const id = await el.getAttribute('id') || '';
    const placeholder = await el.getAttribute('placeholder') || '';
    const value = await el.evaluate(e => e.value || '').catch(() => '');
    console.log(`  ${tag} type=${type} name=${name} id=${id} placeholder=${placeholder} value=${value.substring(0, 40)}`);
  }

  // --- Submit ---
  console.log('\nLooking for submit button...');
  const submitBtn = await page.locator('button[type="submit"], input[type="submit"], button:has-text("Submit"), button:has-text("Apply")').first();
  if (await submitBtn.count() > 0) {
    const btnText = await submitBtn.textContent();
    console.log(`Found submit button: "${btnText}"`);
    await screenshot(page, '07-ready-to-submit');

    console.log('Submitting application...');
    await submitBtn.click();
    await page.waitForTimeout(5000);
    await screenshot(page, '08-after-submit');

    const finalUrl = page.url();
    const finalTitle = await page.title();
    console.log('Final URL:', finalUrl);
    console.log('Final Title:', finalTitle);

    // Check for success
    const pageContent = await page.content();
    const isSuccess = pageContent.toLowerCase().includes('thank you') ||
                      pageContent.toLowerCase().includes('application received') ||
                      pageContent.toLowerCase().includes('successfully') ||
                      finalUrl.includes('confirmation') ||
                      finalUrl.includes('success');

    if (isSuccess) {
      console.log('SUCCESS: Application submitted successfully!');
    } else {
      console.log('UNCERTAIN: Could not confirm submission success. Check screenshot.');
    }

    await screenshot(page, '09-final-state');
  } else {
    console.log('ERROR: Submit button not found!');
    await screenshot(page, '07-no-submit-button');
  }

  await browser.close();
  console.log('Browser closed. Script complete.');
}

run().catch(async (err) => {
  console.error('Script failed:', err.message);
  console.error(err.stack);
  process.exit(1);
});

File: scripts/apply-clickhouse.py
#!/usr/bin/env python3
"""
Apply to ClickHouse - Cloud Software Engineer, Identity and Access Management
Job URL: https://job-boards.greenhouse.io/clickhouse/jobs/5803692004
"""

import os
import json
import time
from datetime import date
from playwright.sync_api import sync_playwright, TimeoutError as PlaywrightTimeout

JOB_URL = "https://job-boards.greenhouse.io/clickhouse/jobs/5803692004"
RESUME_PDF = "/home/user/Agents/profile/Hisham Abboud CV.pdf"
SCREENSHOTS_DIR = "/home/user/Agents/output/screenshots"
COVER_LETTER_PATH = "/home/user/Agents/output/cover-letters/clickhouse-cloud-engineer.md"
APPLICATIONS_JSON = "/home/user/Agents/data/applications.json"

CANDIDATE = {
    "first_name": "Hisham",
    "last_name": "Abboud",
    "email": "hiaham123@hotmail.com",
    "phone": "+31 06 4841 2838",
    "linkedin": "https://linkedin.com/in/hisham-abboud",
    "github": "https://github.com/Hishamabboud",
    "location": "Eindhoven, Netherlands",
}

with open(COVER_LETTER_PATH, "r") as f:
    COVER_LETTER_TEXT = f.read()


def screenshot(page, name):
    path = os.path.join(SCREENSHOTS_DIR, f"clickhouse-{name}.png")
    page.screenshot(path=path, full_page=True)
    print(f"Screenshot saved: {path}")
    return path


def try_fill(page, selectors, value, label="field"):
    for selector in selectors:
        try:
            el = page.locator(selector).first
            if el.count() > 0:
                el.fill(value)
                print(f"Filled {label}: {value[:50]}")
                return True
        except Exception:
            continue
    print(f"WARNING: Could not find {label}")
    return False


def log_all_inputs(page):
    """Log all form inputs for debugging."""
    print("\n--- Form elements found ---")
    elements = page.locator("input, select, textarea").all()
    for el in elements:
        try:
            tag = el.evaluate("e => e.tagName")
            type_ = el.get_attribute("type") or ""
            name = el.get_attribute("name") or ""
            id_ = el.get_attribute("id") or ""
            placeholder = el.get_attribute("placeholder") or ""
            label_for = ""
            # Try to find associated label
            if id_:
                try:
                    label_el = page.locator(f"label[for='{id_}']")
                    if label_el.count() > 0:
                        label_for = label_el.text_content() or ""
                except Exception:
                    pass
            print(f"  {tag} type={type_} name={name} id={id_} placeholder={placeholder} label={label_for.strip()}")
        except Exception as e:
            print(f"  (could not read element: {e})")
    print("--- End form elements ---\n")


def run():
    result = {
        "status": "failed",
        "screenshots": [],
        "notes": "",
    }

    with sync_playwright() as p:
        browser = p.chromium.launch(
            headless=True,
            args=["--no-sandbox", "--disable-setuid-sandbox", "--disable-dev-shm-usage"],
        )
        context = browser.new_context(
            user_agent="Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36"
        )
        page = context.new_page()

        try:
            print(f"Navigating to: {JOB_URL}")
            page.goto(JOB_URL, wait_until="networkidle", timeout=60000)
            result["screenshots"].append(screenshot(page, "01-job-page-loaded"))
            print(f"Page title: {page.title()}")
            print(f"URL: {page.url()}")

            # Wait for form
            try:
                page.wait_for_selector("form", timeout=20000)
                print("Form found on page")
            except PlaywrightTimeout:
                print("No form found within timeout")
                result["notes"] = "No form found on page"
                result["screenshots"].append(screenshot(page, "error-no-form"))
                return result

            result["screenshots"].append(screenshot(page, "02-form-visible"))

            # Log all inputs
            log_all_inputs(page)

            # --- Fill basic fields ---
            try_fill(page,
                ["input#first_name", "input[name='job_application[first_name]']", "input[autocomplete='given-name']"],
                CANDIDATE["first_name"], "first name")

            try_fill(page,
                ["input#last_name", "input[name='job_application[last_name]']", "input[autocomplete='family-name']"],
                CANDIDATE["last_name"], "last name")

            try_fill(page,
                ["input#email", "input[name='job_application[email]']", "input[type='email']"],
                CANDIDATE["email"], "email")

            try_fill(page,
                ["input#phone", "input[name='job_application[phone]']", "input[type='tel']"],
                CANDIDATE["phone"], "phone")

            result["screenshots"].append(screenshot(page, "03-basic-fields-filled"))

            # --- Upload Resume ---
            if os.path.exists(RESUME_PDF):
                file_inputs = page.locator("input[type='file']").all()
                if file_inputs:
                    file_inputs[0].set_input_files(RESUME_PDF)
                    print(f"Resume uploaded: {RESUME_PDF}")
                    page.wait_for_timeout(3000)
                    result["screenshots"].append(screenshot(page, "04-resume-uploaded"))
                else:
                    print("WARNING: No file input found for resume upload")
            else:
                print(f"WARNING: Resume PDF not found: {RESUME_PDF}")

            # --- LinkedIn ---
            try_fill(page,
                ["input#job_application_answers_attributes_0_text_value",
                 "input[id*='linkedin']", "input[name*='linkedin']",
                 "input[placeholder*='LinkedIn']", "input[placeholder*='linkedin']"],
                CANDIDATE["linkedin"], "LinkedIn URL")

            # --- Cover Letter (text area) ---
            cover_letter_selectors = [
                "textarea#cover_letter",
                "textarea[name*='cover_letter']",
                "textarea[id*='cover']",
            ]
            filled_cover = False
            for sel in cover_letter_selectors:
                try:
                    el = page.locator(sel).first
                    if el.count() > 0:
                        el.fill(COVER_LETTER_TEXT)
                        print("Filled cover letter text area")
                        filled_cover = True
                        break
                except Exception:
                    continue
            if not filled_cover:
                print("Cover letter text area not found (may be file upload only)")

            # --- Handle custom questions ---
            # Look for any text inputs/textareas not yet filled
            textareas = page.locator("textarea").all()
            for ta in textareas:
                try:
                    ta_id = ta.get_attribute("id") or ""
                    ta_name = ta.get_attribute("name") or ""
                    current_val = ta.evaluate("e => e.value")
                    if current_val:
                        continue  # already filled
                    # Find label
                    label_text = ""
                    if ta_id:
                        label_el = page.locator(f"label[for='{ta_id}']")
                        if label_el.count() > 0:
                            label_text = label_el.text_content() or ""
                    print(f"Found unfilled textarea: id={ta_id} name={ta_name} label='{label_text.strip()}'")
                    # Fill with relevant content based on label
                    label_lower = label_text.lower()
                    if "cover" in label_lower or "letter" in label_lower:
                        ta.fill(COVER_LETTER_TEXT)
                        print("  -> Filled as cover letter")
                    elif "auth" in label_lower or "iam" in label_lower or "experience" in label_lower:
                        ta.fill("I have experience with OAuth2/OIDC authentication flows, integrating third-party identity providers, and building secure REST APIs with access control patterns in .NET and Python. I am eager to deepen this expertise in SAML, SCIM, and cloud IAM standards.")
                        print("  -> Filled as auth experience")
                except Exception as e:
                    print(f"  (error processing textarea: {e})")

            # --- Handle visa/work authorization selects ---
            selects = page.locator("select").all()
            for sel in selects:
                try:
                    sel_id = sel.get_attribute("id") or ""
                    sel_name = sel.get_attribute("name") or ""
                    label_text = ""
                    if sel_id:
                        label_el = page.locator(f"label[for='{sel_id}']")
                        if label_el.count() > 0:
                            label_text = label_el.text_content() or ""
                    print(f"Select found: id={sel_id} name={sel_name} label='{label_text.strip()}'")
                    label_lower = label_text.lower()
                    id_lower = (sel_id + " " + sel_name).lower()
                    if "visa" in label_lower or "sponsor" in label_lower or "visa" in id_lower or "sponsor" in id_lower:
                        # Candidate is in Netherlands and does not need sponsorship
                        try:
                            sel.select_option(label="No")
                            print(f"  -> Selected 'No' for visa sponsorship")
                        except Exception:
                            try:
                                sel.select_option(value="0")
                                print(f"  -> Selected value 0 for visa sponsorship")
                            except Exception as e2:
                                print(f"  -> Could not set visa select: {e2}")
                    elif "gender" in label_lower or "race" in label_lower or "veteran" in label_lower or "disability" in label_lower:
                        # EEOC - select "Decline to identify" or "I don't wish to answer"
                        options = sel.locator("option").all()
                        option_texts = []
                        for opt in options:
                            option_texts.append(opt.text_content() or "")
                        print(f"  EEOC options: {option_texts}")
                        # Try to find decline option
                        decline_keywords = ["decline", "prefer not", "don't wish", "not answer", "no answer"]
                        for opt_text in option_texts:
                            if any(kw in opt_text.lower() for kw in decline_keywords):
                                try:
                                    sel.select_option(label=opt_text)
                                    print(f"  -> Selected EEOC decline option: {opt_text}")
                                except Exception:
                                    pass
                                break
                except Exception as e:
                    print(f"  (error processing select: {e})")

            result["screenshots"].append(screenshot(page, "05-all-fields-filled"))

            # --- Final check before submit ---
            log_all_inputs(page)
            result["screenshots"].append(screenshot(page, "06-before-submit"))

            # --- Submit ---
            submit_selectors = [
                "input[type='submit']",
                "button[type='submit']",
                "button:has-text('Submit Application')",
                "button:has-text('Submit')",
                "button:has-text('Apply Now')",
                "button:has-text('Apply')",
            ]

            submit_btn = None
            for sel in submit_selectors:
                try:
                    btn = page.locator(sel).first
                    if btn.count() > 0:
                        submit_btn = btn
                        btn_text = btn.text_content() or btn.get_attribute("value") or "Submit"
                        print(f"Found submit button: '{btn_text}'")
                        break
                except Exception:
                    continue

            if submit_btn:
                print("Submitting application...")
                submit_btn.click()
                page.wait_for_timeout(6000)
                result["screenshots"].append(screenshot(page, "07-after-submit"))

                final_url = page.url()
                final_content = page.content().lower()
                print(f"Final URL: {final_url}")

                success_signals = [
                    "thank you" in final_content,
                    "application received" in final_content,
                    "successfully submitted" in final_content,
                    "we've received" in final_content,
                    "confirmation" in final_url,
                    "success" in final_url,
                    "thank" in final_url,
                ]

                if any(success_signals):
                    print("SUCCESS: Application submitted!")
                    result["status"] = "applied"
                    result["notes"] = "Application submitted successfully via Greenhouse form"
                else:
                    # Check for validation errors
                    errors = page.locator(".error, .alert, [class*='error'], [class*='invalid']").all()
                    error_texts = []
                    for err in errors[:5]:
                        try:
                            t = err.text_content()
                            if t and t.strip():
                                error_texts.append(t.strip())
                        except Exception:
                            pass
                    if error_texts:
                        print(f"VALIDATION ERRORS: {error_texts}")
                        result["notes"] = f"Validation errors: {'; '.join(error_texts)}"
                    else:
                        print("UNCERTAIN: Could not confirm submission. Check screenshots.")
                        result["notes"] = "Submitted but could not confirm success from page content"
                        result["status"] = "applied"  # Assume submitted if no errors

                result["screenshots"].append(screenshot(page, "08-final-state"))
            else:
                print("ERROR: Submit button not found!")
                result["notes"] = "Submit button not found"
                result["screenshots"].append(screenshot(page, "error-no-submit"))

        except PlaywrightTimeout as e:
            print(f"TIMEOUT: {e}")
            result["notes"] = f"Timeout: {e}"
            try:
                result["screenshots"].append(screenshot(page, "error-timeout"))
            except Exception:
                pass
        except Exception as e:
            print(f"ERROR: {e}")
            import traceback
            traceback.print_exc()
            result["notes"] = f"Error: {e}"
            try:
                result["screenshots"].append(screenshot(page, "error-exception"))
            except Exception:
                pass
        finally:
            browser.close()

    return result


def update_applications(result):
    """Update the applications.json tracker."""
    with open(APPLICATIONS_JSON, "r") as f:
        apps = json.load(f)

    new_entry = {
        "id": "app-clickhouse-cloud-iam-001",
        "company": "ClickHouse",
        "role": "Cloud Software Engineer - Identity and Access Management",
        "url": JOB_URL,
        "date_applied": str(date.today()),
        "score": 8.0,
        "status": result["status"],
        "resume_file": RESUME_PDF,
        "cover_letter_file": COVER_LETTER_PATH,
        "screenshot": result["screenshots"][-1] if result["screenshots"] else "",
        "all_screenshots": result["screenshots"],
        "notes": result["notes"],
        "response": "",
    }

    apps.append(new_entry)

    with open(APPLICATIONS_JSON, "w") as f:
        json.dump(apps, f, indent=2)

    print(f"\nApplication logged to {APPLICATIONS_JSON}")
    print(f"Status: {result['status']}")
    return new_entry


if __name__ == "__main__":
    print("=" * 60)
    print("ClickHouse - Cloud Software Engineer IAM Application")
    print("=" * 60)
    result = run()
    entry = update_applications(result)
    print("\n--- SUMMARY ---")
    print(json.dumps(entry, indent=2))
